### PR TITLE
Data warehouse foundations

### PR DESCRIPTION
**Pull request by Kevin-Chant**
### Features and Changes

Creates and updates necessary helpers for the `growthbook_clickhouse` datasource type as well as the data enrichment flow.

### CODE REVIEW

File: packages/back-end/src/controllers/datasources.ts [hunk 1]
```diff
@@ -111,7 +111,7 @@ export async function deleteDataSource(
     );
   }
 
-  await deleteDatasourceById(datasource.id, org.id);
+  await deleteDatasourceById(datasource, org.id);
```
Comment (toots) [inline packages/back-end/src/controllers/datasources.ts:114]: This should either reverted or the name should be changed. I don't see why this change slipped through.

File: packages/back-end/src/models/DataSourceModel.ts [hunk 2]
```diff
@@ -88,6 +88,13 @@ export async function getDataSourcesByOrganization(
     context.permissions.canReadMultiProjectResource(ds.projects)
   );
 }
+// WARNING: This does not restrict by organization
+export async function _dangerousGetAllGrowthbookClickhouseDataSources() {
```
Comment (toots) [inline packages/back-end/src/models/DataSourceModel.ts:92]: I thought we wanted to cache this? Maybe at least a in-code cache using a dangling `let` variable.

File: packages/back-end/src/models/DataSourceModel.ts [hunk 3]
```diff
@@ -124,12 +131,15 @@ export async function removeProjectFromDatasources(
   );
 }
 
-export async function deleteDatasourceById(id: string, organization: string) {
+export async function deleteDatasourceById(
```
Comment (toots) [inline packages/back-end/src/models/DataSourceModel.ts:134]: ditto

**Review by toots — COMMENTED** [review]
Could of changes I think we should apply before merging.

File: packages/back-end/src/models/DataSourceModel.ts (hunk 2 above)
Comment (Kevin-Chant) [reply packages/back-end/src/models/DataSourceModel.ts:92]: Yeah that's what I had initially but Jeremy suggested letting the ingestion endpoint handle the caching rather than having two caching layers.

File: packages/back-end/src/api/ingestion/ingestion.router.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    const dataSourcesByOrgId = Object.fromEntries(
+      dataSources.map((ds) => [ds.organization, ds.id])
+    );
+    const sdkConnections = await findSdkConnectionsAcrossMultipleOrgs(
+      Object.keys(dataSourcesByOrgId)
+    );
+    const sdkData = Object.fromEntries(
+      sdkConnections.map((conn) => [
+        conn.id,
+        sdkInfo(conn, dataSourcesByOrgId[conn.organization] || ""),
```
Comment (jdorn) [inline packages/back-end/src/api/ingestion/ingestion.router.ts:50]: Instead of defaulting to an empty string, we should skip these entries and log an error

File: packages/back-end/src/api/ingestion/ingestion.router.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+interface GetDataEnrichmentResponse {
+  sdkData: {
+    [key: string]: SdkInfo;
+  };
+}
+
+function sdkInfo(conn: SDKConnectionInterface, datasource: string): SdkInfo {
+  return {
+    organization: conn.organization,
+    client_key: conn.id,
```
Comment (jdorn) [inline packages/back-end/src/api/ingestion/ingestion.router.ts:26]: This should be `conn.key` instead I believe

File: packages/back-end/src/api/ingestion/ingestion.router.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    const dataSources = await _dangerousGetAllGrowthbookClickhouseDataSources();
+    const dataSourcesByOrgId = Object.fromEntries(
+      dataSources.map((ds) => [ds.organization, ds.id])
+    );
+    const sdkConnections = await findSdkConnectionsAcrossMultipleOrgs(
+      Object.keys(dataSourcesByOrgId)
+    );
+    const sdkData = Object.fromEntries(
+      sdkConnections.map((conn) => [
+        conn.id,
```
Comment (jdorn) [inline packages/back-end/src/api/ingestion/ingestion.router.ts:49]: Same here, this should be `conn.key` instead

File: packages/back-end/src/controllers/datasources.ts [hunk 7]
```diff
@@ -491,7 +491,8 @@ export async function postGoogleOauthRedirect(
   const context = getContextFromReq(req);
   const { projects } = req.body;
 
-  if (!context.permissions.canCreateDataSource({ projects })) {
+  // Huh, why? -- rb
+  if (!context.permissions.canCreateDataSource({ projects, type: undefined })) {
```
Comment (jdorn) [inline packages/back-end/src/controllers/datasources.ts:495]: This function is super old and not actually reachable in the UI anymore, but if we want to be correct here, the type should be `"google_analytics"`

File: packages/back-end/src/models/DataSourceModel.ts (hunk 2 above)
Comment (jdorn) [reply packages/back-end/src/models/DataSourceModel.ts:92]: Yeah, it's using an index so will be super fast.  Dealing with cache invalidation in a horizontally scaled deployment is not fun and not worth it here.

File: packages/back-end/src/models/SdkConnectionModel.ts [hunk 8]
```diff
@@ -132,6 +132,16 @@ export async function findSDKConnectionsByOrganization(
   );
 }
 
+export async function findSdkConnectionsAcrossMultipleOrgs(
```
Comment (jdorn) [inline packages/back-end/src/models/SdkConnectionModel.ts:135]: Should we name this as "dangerous" as well?

File: packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts [hunk 9]
```diff
@@ -181,7 +181,10 @@ export const postDemoDatasourceProject = async (
 
   if (
     !context.permissions.canCreateMetric({ projects: [demoProjId] }) ||
-    !context.permissions.canCreateDataSource({ projects: [demoProjId] })
+    !context.permissions.canCreateDataSource({
+      projects: [demoProjId],
+      type: undefined,
```
Comment (jdorn) [inline packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:186]: The type here should be `"postgres"` since that's what we use for the demo data

File: packages/back-end/src/api/ingestion/ingestion.router.ts (hunk 4 above)
Comment (Kevin-Chant) [reply packages/back-end/src/api/ingestion/ingestion.router.ts:50]: I realized that with some of the changes we made this isn't actually necessary anymore. The query that pulls these `conn`s is by their organization which is just the keys of the object being indexed here, so this should actually be safe now.

File: packages/back-end/src/models/DataSourceModel.ts (hunk 2 above)
Comment (romain-growthbook) [reply packages/back-end/src/models/DataSourceModel.ts:92]: Word.

**Review by romain-growthbook — APPROVED** [review]
Seems good to go to me.
